tritium-metrics/src/main/java/com/palantir/tritium/metrics/tls/InstrumentedSslEngine.java

```diff
@@ -0,0 +1,314 @@
+    public String[] getEnabledCipherSuites() {
+        return engine.getEnabledCipherSuites();
+    }
+
+    @Override
+    public void setEnabledCipherSuites(String[] strings) {
+        engine.setEnabledCipherSuites(strings);
+    }
+
+    @Override
+    public String[] getSupportedProtocols() {
+        return engine.getSupportedProtocols();
+    }
+
+    @Override
+    public String[] getEnabledProtocols() {
+        return engine.getEnabledProtocols();
+    }
+
+    @Override
+    public void setEnabledProtocols(String[] strings) {
+        engine.setEnabledProtocols(strings);
+    }
+
+    @Override
+    public SSLSession getSession() {
+        return engine.getSession();
+    }
+
+    @Override
+    public SSLSession getHandshakeSession() {
+        return engine.getHandshakeSession();
+    }
+
+    @Override
+    public void beginHandshake() throws SSLException {
+        engine.beginHandshake();
+        handshaking.set(true);
+    }
+
+    @Override
+    public SSLEngineResult.HandshakeStatus getHandshakeStatus() {
+        return engine.getHandshakeStatus();
+    }
+
+    @Override
+    public void setUseClientMode(boolean mode) {
+        engine.setUseClientMode(mode);
+    }
+
+    @Override
+    public boolean getUseClientMode() {
+        return engine.getUseClientMode();
+    }
+
+    @Override
+    public void setNeedClientAuth(boolean need) {
+        engine.setNeedClientAuth(need);
+    }
+
+    @Override
+    public boolean getNeedClientAuth() {
+        return engine.getNeedClientAuth();
+    }
+
+    @Override
+    public void setWantClientAuth(boolean want) {
+        engine.setWantClientAuth(want);
+    }
+
+    @Override
+    public boolean getWantClientAuth() {
+        return engine.getWantClientAuth();
+    }
+
+    @Override
+    public void setEnableSessionCreation(boolean flag) {
+        engine.setEnableSessionCreation(flag);
+    }
+
+    @Override
+    public boolean getEnableSessionCreation() {
+        return engine.getEnableSessionCreation();
+    }
+
+    @Override
+    public SSLParameters getSSLParameters() {
+        return engine.getSSLParameters();
+    }
+
+    @Override
+    public void setSSLParameters(SSLParameters sslParameters) {
+        engine.setSSLParameters(sslParameters);
+    }
+
+    // JAVA 9
+
+    @Nullable
+    private static final Method getApplicationProtocol = engineMethodNullable("getApplicationProtocol");
+    @Nullable
+    private static final Method getHandshakeApplicationProtocol =
+            engineMethodNullable("getHandshakeApplicationProtocol");
+    @Nullable
+    private static final Method setHandshakeApplicationProtocolSelector =
+            engineMethodNullable("setHandshakeApplicationProtocolSelector", BiFunction.class);
+    @Nullable
+    private static final Method getHandshakeApplicationProtocolSelector =
+            engineMethodNullable("getHandshakeApplicationProtocolSelector");
+
+    // Override(java9+)
+    public String getApplicationProtocol() {
```

Comment:
We're not the only ones who use reflection check for java 9+ methods. We should provide SslEngine8 and SslEngine9 implementations, only the latter should implement these.